capstone/src/test/java/com/google/sps/servlets/ChallengeServletTest.java

```diff
@@ -0,0 +1,177 @@
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import com.google.sps.Objects.Challenge;
+
+public class ChallengeServletTest {
+
+  private static final String USER_EMAIL = "test@test.com";
+  private static final String USER_ID = "test";
+  private static final String NEW_CHALLENGE = "Bike 20 miles";
+  private static final String CHALLENGE_NAME = "Run";
+  private static final long CHALLENGE_ID = 2;
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+              new LocalDatastoreServiceTestConfig()
+                  .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+              new LocalUserServiceTestConfig())
+          .setEnvEmail(USER_EMAIL)
+          .setEnvIsLoggedIn(true)
+          .setEnvAuthDomain("gmail.com")
+          .setEnvAttributes(
+              new HashMap(
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private ChallengeServlet challengeServlet;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Add test data
+    ImmutableList.Builder<Entity> challenge = ImmutableList.builder();
+    challenge.add(createChallenge(CHALLENGE_NAME));
+    datastore.put(challenge.build());
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    challengeServlet = new ChallengeServlet();
+  }
+
+  /**
+   * Creates challenge entity.
+   *
+   * @param text challenge name
+   * @return entity
+   */
+  private Entity createChallenge(String text) {
+    Entity challengeEntity = new Entity("Challenge");
+    challengeEntity.setProperty("name", text);
+    challengeEntity.setProperty("dueDate", this.getDueDate());
+    challengeEntity.setProperty("votes", new ArrayList<String>());
+    challengeEntity.setProperty("timestamp", System.currentTimeMillis());
+    return challengeEntity;
+  }
+
+  /**
+   * Sets due date to midnight, 7 days from when challenge is posted.
+   *
+   * @return due date in milliseconds
+   */
+  private long getDueDate() {
+    LocalDateTime now = LocalDateTime.now();
+    LocalDateTime dueDate = now.plusDays(7).withHour(23).withMinute(59).withSecond(59).withNano(0);
+    return Timestamp.valueOf(dueDate).getTime();
+  }
```

Comment:
I put this logic in here so that I could check that the due date inputted into the database is calculated correctly. Should I just not assert whether the due dates are equal then?

Reply:
I see. In that case, rather than needing to call the method twice (in createChallenge & the test) you should create the constant variable above and use that throughout.